File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/utils/serialization/ObjectTypeInspector.java
package net.namekdev.entity_tracker.utils.serialization;

import static net.namekdev.entity_tracker.utils.serialization.NetworkSerialization.*;

import java.util.ArrayList;
import java.util.Vector;

import com.artemis.utils.reflect.ClassReflection;
import com.artemis.utils.reflect.Field;

public class ObjectTypeInspector {
	private ArrayList<RegisteredModel> registeredModels = new ArrayList<RegisteredModel>();
	private int lastId = 0;

	private ObjectModelsCollection registeredModelsAsCollection = new ObjectModelsCollection() {
		
		@Override
		public int size() {
			return getRegisteredModelsCount();
		}
		
		@Override
		public ObjectModelNode get(int index) {
			return getRegisteredModelByIndex(index);
		}
		
		@Override
		public ObjectModelNode get(Class<?> type) {
			return inspect(type);
		}

		@Override
		public ObjectModelNode getById(int id) {
			for (RegisteredModel m : registeredModels) {
				if (m.model.id == id)
					return m.model;
			}

			return null;
		}

		@Override
		public void add(ObjectModelNode model) {
			throw new RuntimeException("this implementation shouldn't manually add models. Inspector should do that automatically.");
		}
	};

	
	private static class RegisteredModel {
		public Class<?> type, parentType;
		public ObjectModelNode model;
		
		public RegisteredModel parent;
		public ArrayList<RegisteredModel> children = new ArrayList<>();
	}
	
	public int getRegisteredModelsCount() {
		return registeredModels.size();
	}
	
	public ObjectModelNode getRegisteredModelByIndex(int index) {
		RegisteredModel model = registeredModels.get(index);
		return model != null ? model.model : null;
	}
	
	public ObjectModelNode getModelById(int id) {
		for (RegisteredModel model : registeredModels) {
			if (model.model.id == id) {
				return model.model;
			}
		}
		
		return null;
	}
	
	
	/**
	 * Returns tree description of class type.
	 */
	public ObjectModelNode inspect(Class<?> type) {
		assert(NetworkSerialization.determineType(type) == Type.Unknown);
		
		return inspectLevels(type, null, null, null);
	}

	private ObjectModelNode inspectLevels(Class<?> type, Class<?> parentType, ObjectModelNode parentOfRoot, RegisteredModel parentRegisteredModel) {
		RegisteredModel registeredModel = findModel(type, parentType, parentOfRoot);

		if (registeredModel != null) {
			return registeredModel.model;
		}

		ObjectModelNode root = null;

		if (!type.isArray()) {
			Field[] fields = ClassReflection.getDeclaredFields(type);
	
			ObjectModelNode model = new ObjectModelNode(registeredModelsAsCollection, ++lastId, root);
			model.networkType = Type.Object;
			model.children = new Vector<>(fields.length);
		
			registeredModel = rememberType(type, parentType, model, parentRegisteredModel);
			root = registeredModel.model;
	
			for (Field field : fields) {
				Class<?> fieldType = field.getType();
				ObjectModelNode child = null;
	
				if (fieldType.isArray()) {
					child = inspectArrayType(fieldType, type, registeredModel);
				}
				else {
					Type networkType = NetworkSerialization.determineType(fieldType);
	
					if (networkType == Type.Unknown) {
						RegisteredModel registeredChildModel = findModel(fieldType, type, root);
						
						if (registeredChildModel == null) {
							child = inspectLevels(fieldType, type, root, registeredModel);
						}
						else {
							child = new ObjectModelNode(registeredModelsAsCollection, ++lastId, root).copyFrom(
								registeredChildModel.model
							);
						}
					}
					else if (networkType == Type.Enum) {
						 child = inspectEnum((Class<Enum>) fieldType, type, registeredModel);
					}
					else {
						child = new ObjectModelNode(registeredModelsAsCollection, ++lastId, root);
						child.networkType = networkType;
					}
				}
	
				// TODO because of this we may have to clone what's inside of RegisteredModel
				assert(child.name == null);
				child.name = field.getName();
	
				model.children.addElement(child);
			}
			
			return model;
		}
		else {
			return inspectArrayType(type, parentType, parentRegisteredModel);
		}
	}

	private ObjectModelNode inspectArrayType(Class<?> fieldType, Class<?> parentType, RegisteredModel parentRegisteredModel) {
		ObjectModelNode model = new ObjectModelNode(registeredModelsAsCollection, ++lastId, parentRegisteredModel != null ? parentRegisteredModel.model : null);
		RegisteredModel registeredModel = rememberType(fieldType, parentType, model, parentRegisteredModel);
		
		Class<?> arrayElType = fieldType.getComponentType();
		Type arrayType = determineType(arrayElType);

		
		if (arrayType == Type.Enum) {
			// TODO!
//			throw new RuntimeException("TODO array of enums");
			ObjectModelNode enumFieldModel = inspectEnum((Class<Enum>) arrayElType, fieldType, registeredModel);
			model.children = new Vector<>(1);
			model.children.addElement(enumFieldModel);
		}
		
		// TODO probably that should inspect deeper anyway!
		else if (!isSimpleType(arrayType)) {
//			model = inspectLevels(arrayElType, root);
//
//			if (model.networkType == TYPE_TREE) {
//				arrayType = TYPE_TREE;
//			}
			
			arrayType = arrayElType.isArray() ? Type.Array : Type.Object;
		}

		model.networkType = Type.Array;
		model.childType = (short) arrayType.ordinal();

		return model;
	}
	
	private ObjectModelNode inspectEnum(final Class<Enum> enumType, final Class<?> parentType, final RegisteredModel parentRegisteredModel) {
		// algorithm: will create enum field definition anyway,
		// but first check if there is a need to create a model for enum type (list of possible values) 

		RegisteredModel registeredEnumTypeModel = findModel(enumType, null, null);
		
		if (registeredEnumTypeModel == null) {
			ObjectModelNode enumTypeModel = new ObjectModelNode(registeredModelsAsCollection, ++lastId, null);
			enumTypeModel.networkType = Type.EnumDescription;
			enumTypeModel.name = enumType.getSimpleName();
			
			Enum<?>[] possibleValues = enumType.getEnumConstants();
			enumTypeModel.children = new Vector<>(possibleValues.length);
			registeredEnumTypeModel = rememberType(enumType, null, enumTypeModel, null);
			
			for (int i = 0; i < possibleValues.length; ++i) {
				ObjectModelNode enumValueModel = new ObjectModelNode(registeredModelsAsCollection, ++lastId, enumTypeModel);
				enumValueModel.networkType = Type.EnumValue;
				Enum<?> val = possibleValues[i];
				
				// Note: we cut bytes here, it's not nice but let's believe that no one creates enums greater than 127.
				enumValueModel.childType = (short) val.ordinal();
				enumValueModel.name = val.name();
				enumTypeModel.children.addElement(enumValueModel);
				
				rememberType(null, enumType, enumValueModel, registeredEnumTypeModel);
			}
		}
		
		ObjectModelNode enumFieldModel = new ObjectModelNode(registeredModelsAsCollection, ++lastId, parentRegisteredModel.model);
		enumFieldModel.networkType = Type.Enum;
		
		ObjectModelNode enumModelRef = new ObjectModelNode(registeredModelsAsCollection, registeredEnumTypeModel.model.id, enumFieldModel);
		enumFieldModel.children = new Vector<>(1);
		enumFieldModel.children.addElement(enumModelRef);
		
		rememberType(enumType, parentType, enumFieldModel, parentRegisteredModel);
		
		return enumFieldModel;
	}

	private RegisteredModel findModel(final Class<?> type, final Class<?> parentType, final ObjectModelNode parent) {
		for (RegisteredModel registered : registeredModels) {
			boolean sameParentModel = (parent == null && registered.model.parent == null)
				|| (parent != null && parent.equals(registered.model));

			if (registered.type != null && registered.type.equals(type) || registered.type == null && type == null) {
				boolean isCyclicModel = false; 

				RegisteredModel cur = findChildType(registered, type);
				isCyclicModel = cur != null;

				if (sameParentModel || isCyclicModel) {
					return registered;
				}
			}
		}
		
		return null;
	}
	
	private RegisteredModel findChildType(RegisteredModel registered, final Class<?> type) {
		RegisteredModel cur = registered;
		for (RegisteredModel child : cur.children) {
			if (child.parentType.equals(type)) {
				return child;
			}
			else {
				return findChildType(child, type);
			}
		}
		
		return null;
	}

	private RegisteredModel rememberType(Class<?> type, Class<?> parentType, ObjectModelNode model, RegisteredModel parentRegisteredModel) {
		RegisteredModel newModel = new RegisteredModel();
		newModel.type = type;
		newModel.model = model;
		newModel.parent = parentRegisteredModel;
		newModel.parentType = parentType;
		
		if (parentRegisteredModel != null) {
			parentRegisteredModel.children.add(newModel);
		}

		this.registeredModels.add(newModel);
		return newModel;
	}
	
	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder("[");

		for (RegisteredModel m : this.registeredModels) {
			sb.append("\n{\n  type: " + toString(m.type));
			sb.append("\n  parentType:" + toString(m.parentType));
			sb.append("\n  model:");
			
			if (m.model == null) {
				sb.append(" null\n");
			}
			else {
				sb.append("\n    id: " + m.model.id);
				sb.append("\n    networkType: " + m.model.networkType);
				sb.append("\n    childType: " + m.model.childType);
				sb.append("\n    name: \"" + m.model.name + "\"");
				sb.append("\n    parent:");
				
				if (m.model.parent != null) {
					sb.append(" (id=" + m.model.parent.id + ")");
				}
				else {
					sb.append(" null");
				}
				
				sb.append("\n    children");
				
				if (m.model.children != null) {
					sb.append(" (" + m.model.children.size() + ")");
					
					for (int i = 0, n = m.model.children.size(); i < n; ++i) {
						ObjectModelNode node = m.model.children.get(i);
						sb.append("\n    {\n      id: " + node.id);
						sb.append("\n      name: " + node.name);
						sb.append("\n    }");
					}
				}
				else {
					sb.append(": null");
				}
			}
			
			sb.append("\n}");
		}
		
		sb.append("\n]");
		
		return sb.toString();
	}
	
	private String toString(Object obj) {
		if (obj == null)
			return "null";
		
		return obj.toString();
	}
}

File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/utils/serialization/ObjectTypeInspector.kt
package net.namekdev.entity_tracker.utils.serialization

import net.namekdev.entity_tracker.utils.serialization.NetworkSerialization.*
import net.namekdev.entity_tracker.utils.serialization.NetworkSerialization.Companion.determineType

import java.util.ArrayList
import java.util.Vector

import net.namekdev.entity_tracker.utils.ReflectionUtils

class ObjectTypeInspector {
    private val registeredModels = ArrayList<RegisteredModel>()
    private var lastId = 0

    private val registeredModelsAsCollection = object : ObjectModelsCollection {

        override fun size(): Int {
            return registeredModelsCount
        }

        override fun get(index: Int): ObjectModelNode {
            return getRegisteredModelByIndex(index)
        }

        override fun get(type: Class<*>): ObjectModelNode {
            return inspect(type)
        }

        override fun getById(id: Int): ObjectModelNode? {
            for (m in registeredModels) {
                if (m.model.id == id)
                    return m.model
            }

            return null
        }

        override fun add(model: ObjectModelNode) {
            throw RuntimeException("this implementation shouldn't manually add models. Inspector should do that automatically.")
        }
    }


    private class RegisteredModel {
        var type: Class<*>? = null
        lateinit var model: ObjectModelNode
        var parentType: Class<*>? = null

        var parent: RegisteredModel? = null
        var children = ArrayList<RegisteredModel>()

        override fun toString(): String {
            return model.name ?: "null"
        }
    }

    val registeredModelsCount: Int
        get() = registeredModels.size

    fun getRegisteredModelByIndex(index: Int): ObjectModelNode {
        val model = registeredModels[index]
        return model.model
    }

    fun getModelById(id: Int): ObjectModelNode? {
        for (model in registeredModels) {
            if (model.model.id == id) {
                return model.model
            }
        }

        return null
    }


    /**
     * Returns tree description of class type.
     */
    fun inspect(type: Class<*>): ObjectModelNode {
        val dataType = NetworkSerialization.determineType(type).first
        assert(dataType == DataType.Unknown || dataType == DataType.Enum)

        return inspectLevels(type, null, null, null)
    }

    private fun inspectLevels(type: Class<*>, parentType: Class<*>?, parentOfRoot: ObjectModelNode?, parentRegisteredModel: RegisteredModel?): ObjectModelNode {
        var registeredModel = findModel(type, parentType, parentOfRoot)

        if (registeredModel != null) {
            return registeredModel.model
        }

        var root: ObjectModelNode

        if (!type.isArray) {
            val fields = ReflectionUtils.getDeclaredFields(type)
                .filter { !it.name.startsWith("this$") } // cover hidden field in non-static inner class

            val model = ObjectModelNode(registeredModelsAsCollection, ++lastId, /* TODO: it was: root*/ null)
            model.dataType = DataType.Object
            model.children = Vector<ObjectModelNode>(fields.size)

            registeredModel = rememberType(type, parentType, model, parentRegisteredModel)
            root = registeredModel.model

            for (field in fields) {
                val fieldType = field.type
                var child: ObjectModelNode? = null

                if (fieldType.isArray) {
                    child = inspectArrayType(fieldType, type, registeredModel)
                }
                else {
                    val (dataType, isTypePrimitive) = NetworkSerialization.determineType(fieldType)

                    if (dataType == DataType.Unknown) {
                        val registeredChildModel = findModel(fieldType, type, root)

                        if (registeredChildModel == null) {
                            child = inspectLevels(fieldType, type, root, registeredModel)
                        }
                        else {
                            child = ObjectModelNode(registeredModelsAsCollection, ++lastId, root).copyFrom(
                                registeredChildModel.model
                            )
                            child.name = null

                            // TODO should we remember this model?
                            rememberType(fieldType, type, root, registeredModel)
                        }
                    }
                    else if (dataType == DataType.Enum) {
                        child = inspectEnum(fieldType as Class<Enum<*>>, type, registeredModel)
                    }
                    else {
                        child = ObjectModelNode(registeredModelsAsCollection, ++lastId, root)
                        child.dataType = dataType
                        child.isTypePrimitive = isTypePrimitive
                    }
                }

                // Every object field has is a unique model
                // so make sure we have a new model here! Then give it a name.
                assert(child.name == null)
                child.name = field.name

                model.children!!.addElement(child)
            }

            return model
        }
        else {
            return inspectArrayType(type, parentType, parentRegisteredModel)
        }
    }

    private fun inspectArrayType(fieldType: Class<*>, parentType: Class<*>?, parentRegisteredModel: RegisteredModel?): ObjectModelNode {
        val model = ObjectModelNode(registeredModelsAsCollection, ++lastId, parentRegisteredModel?.model)
        val registeredModel = rememberType(fieldType, parentType, model, parentRegisteredModel)

        val arrayElType = fieldType.componentType
        var (arrayType, isArrayElTypePrimitive) = determineType(arrayElType)


        if (arrayType == DataType.Enum) {
            val enumFieldModel = inspectEnum(arrayElType as Class<Enum<*>>, fieldType, registeredModel)
            model.children = Vector<ObjectModelNode>(1)
            model.children!!.addElement(enumFieldModel)
//            rememberType(arrayElType, fieldType, enumFieldModel, registeredModel)
        }
        else if (arrayElType.isArray) {
            arrayType = DataType.Array
        }
        else if (arrayType == DataType.Unknown) {
            arrayType = DataType.Object
        }

        model.dataType = DataType.Array
        model.dataSubType = arrayType
        model.isSubTypePrimitive = isArrayElTypePrimitive

        if (model.dataSubType == DataType.Array) {
            val submodel = inspectArrayType(arrayElType, fieldType, registeredModel)
            model.children = Vector()
            model.children!!.addElement(submodel)
        }

        return model
    }

    private fun inspectEnum(enumType: Class<Enum<*>>, parentType: Class<*>, parentRegisteredModel: RegisteredModel): ObjectModelNode {
        // algorithm: always create enum field definition,
        // but first check if there is a need to create
        // a model for enum type (with list of possible values)

        var registeredEnumTypeModel = findModel(enumType, null, null)

        if (registeredEnumTypeModel == null) {
            val enumTypeModel = ObjectModelNode(registeredModelsAsCollection, ++lastId, null)
            enumTypeModel.dataType = DataType.EnumDescription
            enumTypeModel.name = enumType.simpleName

            val possibleValues = enumType.enumConstants
            enumTypeModel.children = Vector<ObjectModelNode>(possibleValues.size)
            registeredEnumTypeModel = rememberType(enumType, null, enumTypeModel, null)

            for (i in possibleValues.indices) {
                val value = possibleValues[i]
                val enumValueModel = ObjectModelNode(registeredModelsAsCollection, ++lastId, enumTypeModel)
                enumValueModel.dataType = DataType.EnumValue
                enumValueModel.enumValue = value.ordinal
                enumValueModel.name = value.name
                enumTypeModel.children!!.addElement(enumValueModel)

                rememberType(null, enumType, enumValueModel, registeredEnumTypeModel)
            }
        }

        val enumFieldModel = ObjectModelNode(registeredModelsAsCollection, ++lastId, parentRegisteredModel.model)
        enumFieldModel.dataType = DataType.Enum
        enumFieldModel.children = Vector<ObjectModelNode>(1)
        enumFieldModel.children!!.addElement(registeredEnumTypeModel.model)

        rememberType(enumType, parentType, enumFieldModel, parentRegisteredModel)

        return enumFieldModel
    }

    private fun findModel(type: Class<*>?, parentType: Class<*>?, parent: ObjectModelNode?): RegisteredModel? {
        for (registered in registeredModels) {
            val sameParentModel = parent === null && registered.model.parent === null || parent !== null && parent === registered.model

            // a != null && a == b || a == null && b == null

            if (registered.type != null && registered.type == type || registered.type == null && type == null) {
                var isCyclicModel = false

                val cur = findChildType(registered, type!!)
                isCyclicModel = cur != null

                if (sameParentModel || isCyclicModel) {
                    return registered
                }
            }
        }

        return null
    }

    private fun findChildType(registered: RegisteredModel, type: Class<*>): RegisteredModel? {
        val cur = registered
        for (child in cur.children) {
            if (child.parentType == type) {
                return child
            }
            else {
                return findChildType(child, type)
            }
        }

        return null
    }

    private fun rememberType(type: Class<*>?, parentType: Class<*>?, model: ObjectModelNode, parentRegisteredModel: RegisteredModel?): RegisteredModel {
        val found = this.registeredModels.find { m-> m.model.id == model.id }

        if (found != null) {
            return found
        }

        val newModel = RegisteredModel()
        newModel.type = type
        newModel.model = model
        newModel.parent = parentRegisteredModel
        newModel.parentType = parentType

        if (parentRegisteredModel != null) {
            parentRegisteredModel.children.add(newModel)
        }

        this.registeredModels.add(newModel)
        return newModel
    }

    override fun toString(): String {
        val sb = StringBuilder("[")

        for (m in this.registeredModels) {
            sb.append("\n{\n  type: " + toString(m.type))
            sb.append("\n  parentType:" + toString(m.parentType))
            sb.append("\n  model:")

            if (m.model == null) {
                sb.append(" null\n")
            }
            else {
                sb.append("\n    id: " + m.model.id)
                sb.append("\n    dataType: " + m.model.dataType)
                sb.append("\n    dataSubType: " + m.model.dataSubType)
                sb.append("\n    name: \"" + m.model.name + "\"")
                sb.append("\n    parent:")

                if (m.model.parent != null) {
                    sb.append(" (id=" + m.model.parent!!.id + ")")
                }
                else {
                    sb.append(" null")
                }

                sb.append("\n    children")

                if (m.model.children != null) {
                    sb.append(" (" + m.model.children!!.size + ")")

                    var i = 0
                    val n = m.model.children!!.size
                    while (i < n) {
                        val node = m.model.children!![i]
                        sb.append("\n    {\n      id: " + node.id)
                        sb.append("\n      name: " + node.name!!)
                        sb.append("\n    }")
                        ++i
                    }
                }
                else {
                    sb.append(": null")
                }
            }

            sb.append("\n}")
        }

        sb.append("\n]")

        return sb.toString()
    }

    private fun toString(obj: Any?): String {
        if (obj == null)
            return "null"

        return obj.toString()
    }
}
